### ISSUE TITLE
Support calls

### ISSUE THREAD

**Issue by soermejo** [LonamiWebs/Telethon#177]
How can i send an mp3 audio when accepting calls?

**Comment by Lonami** [comment]
> when accepting calls

Telegram calls? I have never used the Telegram API through Telethon, so I have no clue how that works. But I guess, instead using a microphone, you would stream the `.mp3` file.

**Comment by seby2000** [comment]
I have no idea how to stream it :( 

_Sent from my Samsung SM-G530FZ using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Lonami** [comment]
Could you explain how you currently accept a call?

**Comment by seby2000** [comment]
Im currently using another client but i think u will receive an update ["UpdatePhoneCall"](https://lonamiwebs.github.io/Telethon/constructors/update_phone_call.html). Then u call ["acceptCallRequest"](https://lonamiwebs.github.io/Telethon/methods/phone/accept_call.html)

**Comment by Lonami** [comment]
But Telethon doesn't even support the calls protocol yet so I have no idea how you pretend to "send" (this should be "stream") an `.mp3` audio!

**Comment by seby2000** [comment]
Do u know on messages.getDhConfig what i have to pass to version and random_lenght parameters?

**Comment by cher-nov** [comment]
@seby2000,
https://core.telegram.org/method/messages.getDhConfig

**Comment by seby2000** [comment]
@cher-nov i have already read it. 
"Value of the version parameter from messages.dhConfig, avialable at the client"
What does it mean?
"Length of the required random sequence"
This too...

**Comment by MasterGroosha** [comment]
So, is it possible to initiate calls or not?

**Comment by Lonami** [comment]
Not sure, I haven't looked into it. If all you need to make the calls is use the Telegram API, then sure you can since all methods are available. If some other weird encryption is required, or since it's peer to peer I think?, then no, there's nothing built for that on Telethon yet. Maybe interfacing [libtgvoip](https://github.com/telegramdesktop/libtgvoip) is enough.

**Comment by danog** [comment]
> I read this lib but i didnt understand much https://github.com/danog/libtgvoip

Try https://github.com/danog/MadelineProto/blob/master/magna.php 🗡 

P.S. Sorry Lonami had to do it <3

**Comment by Lonami** [comment]
@danog you know you can also contribute to Telethon right? It's not that hard, and Python is easy to read :)

Even though Madeline is indeed great and has been added to the list of [Telegram apps](https://telegram.org/apps) some people may still prefer to go with an implementation written in Python itself, if Python is what they use, since they'll depend on no other external tools.

**Comment by stek29** [comment]
@Lonami AES IGE without openssl when?

**Comment by giuseppeM99** [comment]
Daniil is writing madeline.py, but he will release it only once he'll get 50 donation of 1[€$whatever] or more....

**Comment by Lonami** [comment]
> @Lonami AES IGE without openssl when?

@stek29 that has been a thing since the very beginning. Making use of openssl if available (see issue 199) was a recent enhancement.

> but he will release it only once he'll get 50 donation

Of course, because PHP is his first priority, and he isn't going to spend time for free like that. Telethon on the other hand is just Python alone, and another open source project on its own which I develop only because I want to.

**Comment by devxpy** [comment]
@Lonami, how hard would it be to do a python wrapper for libtgvoip? I am ready to contribute if it takes a reasonable amount of time.
I am working on a project where I need to do voice communications. I was doing voice messages since they are very reliable but I tried out telegram calls and they are just fabulous. I've never experienced such perfect call quality on the internet, on a very shitty internet connection

**Comment by Lonami** [comment]
It would not be excessively hard, since there's already a [`Telethon-calls`](https://github.com/LonamiWebs/Telethon-calls) proof of concept. It only needs said considerable amount of work to integrate it nicely and not recompile code every time one makes a call ;)

**Comment by Lonami** [comment]
If you really are willing to work on it feel free to fork the project and maintain the extension yourself. I'll help as much as I can, but calls are definitely not a priority on my list.

**Comment by devxpy** [comment]
Can you please elaborate on what exactly are the problems with that proof-of-concept as of now?
Also, if possible a small intro on what i need to get that working.

I also have a more feasible solution, by using [zeromq](http://zeromq.org/)
Its a fantastic way to "Connect your code in any language, on any platform."
I have personally used it for a project before and it works like a charm.

Not as perfect as a binding but definitely easier to implement.

**Comment by Lonami** [comment]
> Can you please elaborate on what exactly are the problems with that proof-of-concept as of now?

It's calling a external compiler to generate the code that embeds the required parameters to stabilish the call ;)

**Comment by devxpy** [comment]
> There isn't really a need for zeromq. A C extension (like cryptg, really small) should be able to do the job.

I was actually looking into using zeromq on php-libtgvoip. 

I am partial to zeromq since its a less-headache solution overall.

I also felt that the libtgvoip is very poorly documented and writing zeromq communication channel would be much easier than reverse-engineering their whole API

> It's calling a external compiler to generate the code that embeds the required parameters to stabilish the call ;)
that sounds bad

**Comment by Lonami** [comment]
> I also felt that the libtgvoip is very poorly documented

All the "hard" work is already done on that repository I keep linking. It's only missing a few tweaks. Shouldn't be too hard to pick up. C extensions for Python aren't that hard either really (also linked a small one you can use as a base).

**Comment by devxpy** [comment]
I forgot to ask, does this run on telegram servers directly?

**Comment by Lonami** [comment]
No? Calls are peer to peer.

**Comment by devxpy** [comment]
WHAT. okay, now i can understand why the quality is so great, not intermediary to deal with :+1:

**Comment by Lonami** [comment]
I'm not sure if there's an option to disable peer to peer. Don't take my word on this.

**Comment by devxpy** [comment]
I dont want to disable it. I like it being p2p rather than having a server to deal with.

**Comment by devxpy** [comment]
Did you create `DynamicDict` just so that you wont have to write `[' ']` ?

**Comment by Lonami** [comment]
Yes. `['']` is annoying to type. We can move this conversation to [Telegram](https://t.me/Lonami) if you prefer.
